### PR TITLE
Hw26

### PR DESCRIPTION
**Pull request by frostpv**

### CODE REVIEW

File: src/main/java/com/internet/shop/controllers/CartController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+public class CartController extends HttpServlet {
+    private static final Long USER_ID = 1L;
+
+    private static final Injector injector =
+            Injector.getInstance("com.internet.shop");
+
+    ShoppingCartService shoppingCartService =
+            (ShoppingCartService) injector.getInstance(ShoppingCartService.class);
+    ProductService productService =
```
Comment (dariiabutorina) [inline src/main/java/com/internet/shop/controllers/CartController.java]: Access modifiers

File: src/main/java/com/internet/shop/controllers/ProductAddController.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+        req.getRequestDispatcher("/WEB-INF/views/product/add.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String price = req.getParameter("price");
+        if (name.length() >= 1 && price.length() >= 1) {
```
Comment (dariiabutorina) [inline src/main/java/com/internet/shop/controllers/ProductAddController.java]: Check your input data, you may receive the empty Strings

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+
+        String name = req.getParameter("name");
+        String login = req.getParameter("login");
+        String psw = req.getParameter("psw");
+        String pswRepeat = req.getParameter("psw-repeat");
```
Comment (dariiabutorina) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java:32]: Same

**Review by dariiabutorina — CHANGES_REQUESTED** [review]
User - delete, product - add to cart, product - remove from cart methods required

File: src/main/java/com/internet/shop/controllers/CartController.java (hunk 1 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/CartController.java]: added

File: pom.xml [hunk 4]
```diff
@@ -21,12 +21,17 @@
...
-                <groupId>javax.servlet</groupId>
-                <artifactId>javax.servlet-api</artifactId>
-                <version>3.1.0</version>
-                <!--<scope>provided</scope> -->
-            </dependency>
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>4.0.1</version>
+            <!--<scope>provided</scope> -->
```
Comment (Sofasmile) [inline pom.xml]: let's remove commented code from RP

File: src/main/java/com/internet/shop/controllers/CartController.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class CartController extends HttpServlet {
+    private static final Long USER_ID = 1L;
+
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/CartController.java]: let's remove empty lines between fields

File: src/main/java/com/internet/shop/controllers/CartController.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            (ProductService) injector.getInstance(ProductService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        try {
+            ShoppingCart shoppingCart = shoppingCartService.getByUserId(USER_ID);
+            req.setAttribute("cart", shoppingCart.getProducts());
+            req.getRequestDispatcher("/WEB-INF/views/cart/cart.jsp").forward(req, resp);
+        } catch (NoSuchElementException e) {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/CartController.java]: do we need to use catch block here? let's remove it

File: src/main/java/com/internet/shop/controllers/CartController.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            ShoppingCart shoppingCart = new ShoppingCart();
+            shoppingCart.setUserId(USER_ID);
+            shoppingCartService.create(shoppingCart);
+            req.setAttribute("cart", shoppingCart.getProducts());
+            req.getRequestDispatcher("/WEB-INF/views/cart/cart.jsp").forward(req, resp);
+        }
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/CartController.java]: let's create a separate controller for adding products

File: src/main/java/com/internet/shop/controllers/ProductAddController.java (hunk 2 above)
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/ProductAddController.java]: better to do it at the jsp using HTML tags - https://www.w3schools.com/html/html_form_input_types.asp (Input Restrictions, Input Type Range)

File: src/main/java/com/internet/shop/controllers/ProductAddController.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+public class ProductAddController extends HttpServlet {
+    private static final Injector injector =
+            Injector.getInstance("com.internet.shop");
+    private ProductService productService =
+            (ProductService) injector.getInstance(ProductService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/ProductAddController.java]: remove empty line

File: src/main/java/com/internet/shop/controllers/ProductListController.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            Injector.getInstance("com.internet.shop");
+    private ProductService productService =
+            (ProductService) injector.getInstance(ProductService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        List<Product> products = productService.getAll();
+        req.setAttribute("products", products);
+
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/ProductListController.java]: let's remove empty line

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/user/register.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: the same about empty line

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        String name = req.getParameter("name");
+        String login = req.getParameter("login");
+        String psw = req.getParameter("psw");
+        String pswRepeat = req.getParameter("psw-repeat");
+
+        if (psw.equals(pswRepeat)) {
+            User user = new User(name, login, psw);
+            userService.create(user);
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: let's create a shopping cart for the user during registration

File: src/main/java/com/internet/shop/controllers/UsersController.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import com.internet.shop.models.User;
+import com.internet.shop.services.UserService;
+import java.io.IOException;
+import java.util.List;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class UsersController extends HttpServlet {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/UsersController.java]:
```suggestion
public class DeleteUserController extends HttpServlet {
```

File: src/main/java/com/internet/shop/controllers/UsersController.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    private static final Injector injector = Injector.getInstance("com.internet.shop");
+    private UserService userService = (UserService) injector.getInstance(UserService.class);
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        List<User> users = userService.getAll();
+        req.setAttribute("users", users);
+        req.getRequestDispatcher("/WEB-INF/views/user/all.jsp").forward(req, resp);
+    }
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/UsersController.java]: let's move this method to a new controller: GetAllUsersController

File: src/main/java/com/internet/shop/controllers/UsersController.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String id = req.getParameter("id");
+        userService.delete(Long.parseLong(id));
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/UsersController.java]: remove a shopping cart with user

File: pom.xml (hunk 4 above)
Comment (frostpv) [reply pom.xml]: removed

File: src/main/java/com/internet/shop/controllers/CartController.java (hunk 5 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/CartController.java]: removed

File: src/main/java/com/internet/shop/controllers/ProductAddController.java (hunk 8 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/ProductAddController.java]: fixed

File: src/main/java/com/internet/shop/controllers/ProductListController.java (hunk 9 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/ProductListController.java]: removed

File: src/main/java/com/internet/shop/controllers/RegistrationController.java (hunk 10 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/RegistrationController.java]: removed

File: src/main/java/com/internet/shop/controllers/RegistrationController.java (hunk 11 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/RegistrationController.java]: added

File: src/main/java/com/internet/shop/controllers/UsersController.java (hunk 13 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/UsersController.java]: moved

File: src/main/java/com/internet/shop/controllers/UsersController.java (hunk 12 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/UsersController.java]: fixed

File: src/main/java/com/internet/shop/controllers/UsersController.java (hunk 14 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/UsersController.java]: added

File: src/main/java/com/internet/shop/controllers/CartController.java (hunk 6 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/CartController.java]: fixed

File: src/main/java/com/internet/shop/controllers/ProductAddController.java (hunk 2 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/ProductAddController.java]: fixed

File: src/main/java/com/internet/shop/controllers/CartController.java (hunk 7 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/CartController.java]: fixed

File: src/main/java/com/internet/shop/controllers/AddProductInToCartController.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+import com.internet.shop.models.ShoppingCart;
+import com.internet.shop.services.ProductService;
+import com.internet.shop.services.ShoppingCartService;
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AddProductInToCartController extends HttpServlet {
```
Comment (Sofasmile) [inline src/main/java/com/internet/shop/controllers/AddProductInToCartController.java]:
```suggestion
public class AddProductToCartController extends HttpServlet {
```

**Review by Sofasmile — APPROVED** [review]

File: src/main/java/com/internet/shop/controllers/AddProductInToCartController.java (hunk 15 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/AddProductInToCartController.java]: fixed

**Pull request merged**